05_assessing_code_quality/1_project_structure.md

```diff
@@ -0,0 +1,111 @@
+
+For example, if is a data science project, I would expect to see folders for data, notebooks, training models and reports. If it's a web application, I would expect to see folders for templates, static files, and middleware.
+
+Take a moment to look at the project you are working on. Can you tell what it does just by looking at the names of the scripts and folders?
+
+### 2. Make things easy to find
+
+Imagine you're in your sprint planning. You've been tasked with fixing a bug. "It's a simple bug", the senior developer says. "All you have to do is change the `calculate_date` function".
+
+But where is that? You look at the code and you can't find it. You ask the senior developer and he says "It's in the `utils` folder". But there are 20 files in the `utils` folder. Which one is it?
+
+The problem about a folder with a name such as `utils` is that it's too generic and they don't add any semantic meaning to the functions and classes. It's a dumping ground for code that doesn't fit anywhere else. It's okay if you keep them small, but be careful with folders with generic names like `utils`, `functions`, `helper` and `common`: they can quickly become a black hole.
+
+Another problem is the depth of nested folders. The closer a function is to the surface, the easier it is to find. So keep the most important functions and classes close to the surface. That why, in `pandas` you can import the `read_csv` function directly from the `pandas` module. Imagine if you had to import it from `pandas.io.parsers.read_csv`!
+
+### 3. Make things modular
+
+Remember when we said programmers see each new requirement as a new puzzle piece that they must be fit into a puzzle that grow more and more complex?
+
+The solution is to make the puzzle pieces interchangeable. If your puzzle is make only of square pieces, then you can easily swap them around. But if your puzzle is made of pieces of different shapes, then it's considerable harder.
+
+The same applies to your code. If you can swap parts of your code with ease, then your code is modular.
+
+A good strategy is to make your code modular is to think about it as LEGO blocks. Each block should be self-contained and have a single responsibility. Then you have an "entrypoint" (i.e. the main script) that connects the blocks together.
+
+#### What is a "responsibility"?
+
+A responsibility is a reason to change.
+
+As an example, consider a function/module/class that compiles and prints a report. This piece of code can be changed for two reasons:
+
+1. The content of the report could change.
+2. The format of the report could change.
+
+That means that it currently contain two separate responsibilities. Because they are together, if there is a change to the report compilation process, there is a greater danger that the printing code will break. And vice-versa.
```

Comment:
[change] I don't think this is very clear. We have a module that does two actions and has two ways of changing. It is not clear which of these are the responsibilities. 
"two separate responsibilities. Because they are together" is not very congruent. 
It is not clear from the text why breaking the compilation should mean an increased danger of breaking the printing code.

Reply:
O que achas desta alternativa?

```suggestion
That means that it currently contain two separate responsibilities. Because they are together, they may share some state or variables - this means the responsibilities are **coupled** by this shared state.

In our puzzle metaphor, two coupled components are like two puzzle pieces with tabs so unique that can only connect with each other,

Imagine this scenario:  you need to change the order of the report section;  but that caused a variable that was also being used to create the content to change therefore breaking how the content was created. The closer 2 responsibilities are to one another the higher are the chances for that to happen.
```

Reply:
está melhor. ainda assim acho que dizer "separate" seguido de "together" é um pouco confuso. eu apagava a palavra "separate".